### ISSUE TITLE
Finetuning code

### ISSUE THREAD

**Issue by StrangeTcy** [kaiokendev/cutoff-len-is-context-len#2]
My current guess is that any fastchat-based code would work: create the model, apply all the patching you need, and then train? 
The only question I have is about the dataset: does it have to contain long sequences only?

**Comment by kaiokendev** [comment]
It depends.

For linear interpolation, it is possible to use inference well beyond the fine-tuned length, but there is a cost in form of some worse performance. The recommended formula is pre-training length / fine-tuning length, so for example 0.25 for fine-tuning length of 8192. In this case, you can go to 0.125 (16384) even with good performance, but beyond that it will start to break down. But this only concerns linear interpolation. 

For [NTK](https://github.com/jquesnelle/scaled-rope/blob/master/scaled_rope/LlamaPartNTKScaledRotaryEmbedding.py) you should be able to use higher scales (which provide longer maximum inference length) with good performance, regardless of what scale was used during training. But the only thing is that so far I have only conducted perplexity test on this approach. I have not done a full benchmark on NTK performance so I do not know what the performance is like - subjectively it felt different than the linear interpolation but I have not yet confirmed it with numbers.

**Comment by StrangeTcy** [comment]
Oh, sorry. I meant the XPos/Length-Extrapolatable transformers trick.
The testing notebook says:
>The extrapolation issue can be fixed by using the implementation described in https://arxiv.org/abs/2212.10554v1 ("A Length-Extrapolatable Transformer") and unlike ALiBi is compatible with RoPE (as it is a generalization of the latter)
and then
>Any model trained with > 2048 cutoff should be able to print the correct password from above. Due to the stability issue in RoPE, it may need priming, but it can definitely get the password

TL;DR you should train with higher cutoff if you have the data for it

And it's the "the data for it" part I'm concerned about; should we just restructure something like redpajama, or take some summarization dataset, or use something else entirely?

**Comment by kaiokendev** [comment]
It depends on the nature of the dataset. For something like wikitext for example you can just keep appending splits until it suits your needs, or you can use Arxiv dataset if you need consistently long samples. As far as XPos, I think the dataset you use for uptraining does not matter as long as you finetune afterwards.

**Comment by StrangeTcy** [comment]
@kaiokendev 
>For [NTK](https://github.com/jquesnelle/scaled-rope/blob/master/scaled_rope/LlamaPartNTKScaledRotaryEmbedding.py?rgh-link-date=2023-07-13T21%3A06%3A47Z) you should be able to use higher scales (which provide longer maximum inference length) with good performance, regardless of what scale was used during training.

Actually, if you have the training/finetuning args for that NTK `finetune.py` script, that'd be much appreciated. My guess is you've experimented with it a bit.
